### ISSUE TITLE
Make `ServerConfig.blockingTaskExecutor` context-aware

### ISSUE THREAD

**Issue by trustin** [line/armeria#46]
We currently use `Executor` as the type of `ServerConfig.blockingTaskExecutor` property. Although `Executor` provides a nice abstraction for executing long-running tasks to us, we may want more advanced behavior, such as:
- to isolate certain type of invocations and tasks to a different thread pool.
- to execute certain type of invocations and tasks directly (no pooling)

To achieve this, we need to define a new interface similar to `Executor`. To sketch a little bit:

``` java
public interface BlockingTaskExecutor {
    void execute(ServiceInvocationContext ctx, Runnable task);
    void scheduleAtFixedDelay(ServiceInvocationContext ctx, ...);
}
```

We also need to provide a way to compose `BlockingTaskExecutor`s, like we do for `Service`s with `PathMapping`. In this case, we need something more than just doing a path-matching. e.g. hostname, method name, and remote address

**Comment by trustin** [comment]
@anuraaga @inch772 - While working on this issue, I found that I made the following change in `ServerBuilder.java` when we open-source Armeria:

```
     public Server build() {
         Executor blockingTaskExecutor = this.blockingTaskExecutor;
         if (blockingTaskExecutor == null) {
-            blockingTaskExecutor = new ThreadPoolExecutor(
-                    0, DEFAULT_MAX_BLOCKING_TASK_THREADS,
-                    60, TimeUnit.SECONDS, new LinkedTransferQueue<>(),
-                    new DefaultThreadFactory("armeria-blocking-tasks", true));
+            blockingTaskExecutor = defaultBlockingTaskExecutor();
         }
         ...
```

Do you remember why we made this change? Looking again, it seems to me that a new thread pool should be created when a user did not specify `blockingTaskExecutor` and `Server.stop()` should terminate the thread pool when the server stops. I don't see a reason we should use a static daemon thread pool.

**Comment by trustin** [comment]
Beside my question above, I was trying to figure out what the best way of letting a user use a different blocking task executor for a different service. We are currently using the following decorator, which seems to work well:

```java
Executor alternativeExecutor = ...;
Function<ServiceInvocationHandler, ServiceInvocationHandler> alternativeExecutorHandler =
                delegate -> (ctx, unused, promise) -> delegate.invoke(ctx, alternativeExecutor, promise);

server.serviceAt(..., myService.decorate(alternativeExecutorHandler));
```

.. and I'm not very convinced yet if it's a good idea to introduce a new interface like what's proposed in the issue description.

Probably the only viable advantage of the new interface is that we can enforce the life cycle of the executors to be a part of the life cycle of a server. i.e. When we stop a server, we can make sure all executors are terminated.  However, I doubt users always want this behavior; they might be running Armeria in some other managed environment (Spring? ;-)) and want the executors managed by it.

So... we probably should never shutdown/terminate a blockingTaskExecutor unless `ServerBuilder` created it by itself because a user did not specify one... which is actually quite similar to the current behavior.

@anuraaga please let me know if you have any good reason to introduce a new dedicated interface to solve this issue. Otherwise, let me only:

- change the default blockingTaskExecutor behavior and
- add a new decorator method like: `Service.withBlockingTaskExecutor(Executor)`

WDYT?

**Comment by anuraaga** [comment]
I agree it's better not to use a static thread pool - I think I commented on it once but you were ok with it since the threads are created lazily.

And formalizing the blocking executor interface as your proposed decorator sounds fine to me. The implementation details don't matter so much, but coming up with that arcane decorator wasn't easy, and I think it's worth providing users with a simpler interface. I think the most normal case is just one service on a port, but in the less normal case of multiple services there are probably few cases you wouldn't want them on independent executors for safety.

**Comment by minwoox** [comment]
I think this was already implemented, so let me close this issue.
